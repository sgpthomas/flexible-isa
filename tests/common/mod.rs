use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
