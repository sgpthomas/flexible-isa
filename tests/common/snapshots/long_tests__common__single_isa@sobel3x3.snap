---
source: tests/common/mod.rs
description: generate isa for sobel3x3
---
[
    (
        19,
        1,
        "(+ (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        20,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        24,
        1,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> ?#2))) (cast<int32> (- (cast<int32> 0) (cast<int32> (* (cast<int32> ?#3) (cast<int32> ?#4))))))",
    ),
    (
        26,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        27,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<unknown> (* (cast<int32> (+ (cast<int32> (- (cast<int32> (- (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#1) (cast<int32> 4))) (cast<int32> ?#2))) (cast<int32> (* (cast<int32> ?#3) (cast<int32> 4))))) (cast<int32> 1))) (cast<int32> (+ (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#4) (cast<int32> 4))) (cast<int32> ?#5))) (cast<int32> (* (cast<int32> ?#6) (cast<int32> 4))))) (cast<int32> 6))))) (cast<int32> ?#7))) (cast<int32> (* (cast<int32> (+ (cast<int32> ?#8) (cast<int32> ?#9))) (cast<int32> 4))))) (cast<int32> ?#10))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#11) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))))",
    ),
    (
        33,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#3) (cast<int32> ?#4))) (cast<int32> 4))) (cast<int32> ?#5))) (cast<int32> ?x0))) (cast<int32> ?#6))) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        34,
        1,
        "(+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        42,
        1,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) ?x2)) (cast<int32> ?x3))",
    ),
    (
        48,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        50,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 5))",
    ),
    (
        51,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 8))",
    ),
    (
        52,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 13))",
    ),
    (
        75,
        1,
        "(+ (cast<unknown> (* (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> (call[max] (cast<int32> (/ (cast<int32> (+ (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?#2) (cast<int32> ?#3))) (cast<unknown> (+ (cast<unknown> ?#4) (cast<int32> 254))))) (cast<int32> (neg (cast<int32> 127))))) (cast<int32> 128))) (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (/ (cast<int32> (- (cast<int32> 0) (cast<int32> ?#5))) (cast<int32> 128))) (cast<int32> (neg (cast<int32> 1))))) (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#6) (cast<int32> 127))) (cast<int32> 128))))))))) (cast<int32> 128))) ?x0)",
    ),
    (
        89,
        1,
        "(+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<unknown> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> ?x2))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#3) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))))",
    ),
    (
        115,
        1,
        "(+ ?x0 (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x1) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#2))))))",
    ),
    (
        126,
        1,
        "(- (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        127,
        1,
        "(- (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        135,
        1,
        "(- (cast<unknown> ?x0) (cast<int32> (call[max] (cast<int32> ?x1) (cast<int32> ?x2))))",
    ),
    (
        137,
        1,
        "(- ?x0 (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> ?x3))))",
    ),
    (
        142,
        1,
        "(* (cast<int32> (+ (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?#2))) (cast<int32> 1))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#3))",
    ),
    (
        156,
        1,
        "(* (cast<unknown> ?x0) (cast<int32> ?x1))",
    ),
    (
        174,
        1,
        "(<= ?x0 (cast<int32> ?x1))",
    ),
    (
        176,
        1,
        "(== ?#0 ?#1)",
    ),
    (
        177,
        1,
        "(&& ?#0 ?#1)",
    ),
    (
        181,
        1,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        182,
        1,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 130))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        189,
        1,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 512))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        190,
        1,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 514))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        192,
        1,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> ?x0) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        193,
        1,
        "(get ?#3 (cast<unknown> (+ (cast<unknownx128> (call[max] (cast<unknownx128> (call[min] (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?#4))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#5) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> (+ (cast<int32> ?x2) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> ?#6))))))))",
    ),
    (
        5,
        2,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?#2))) (cast<int32> ?x1))",
    ),
    (
        15,
        2,
        "(+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 127))",
    ),
    (
        23,
        2,
        "(+ (cast<int32> (* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (/ (cast<int32> (- (cast<int32> 0) (cast<int32> ?#2))) (cast<int32> 128))) (cast<int32> (neg (cast<int32> 1))))) (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#3) (cast<int32> 127))) (cast<int32> 128))))))) (cast<int32> 128))) ?x0)",
    ),
    (
        36,
        2,
        "(+ (cast<int32> (call[max] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        58,
        2,
        "(+ (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?#1) (cast<int32> 4))))",
    ),
    (
        61,
        2,
        "(+ (cast<int32> ?x0) ?x1)",
    ),
    (
        94,
        2,
        "(+ (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))",
    ),
    (
        112,
        2,
        "(+ ?x0 (cast<int32> (call[max] (cast<int32> ?x1) (cast<int32> ?x2))))",
    ),
    (
        123,
        2,
        "(- (cast<int32> 0) (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))))",
    ),
    (
        129,
        2,
        "(- (cast<int32> (call[min] (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        133,
        2,
        "(- (cast<int32> ?x0) (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> ?x1) (cast<int32> ?x2))) ?x3)))",
    ),
    (
        136,
        2,
        "(- ?x0 (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> ?x1) (cast<int32> ?x2))) (cast<int32> 1))))",
    ),
    (
        146,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<unknown> (+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#2) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> 128))) (cast<int32> 128))))",
    ),
    (
        155,
        2,
        "(* (cast<unknown> ?x0) (cast<int32> 128))",
    ),
    (
        163,
        2,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        166,
        2,
        "(/ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        179,
        2,
        "(get ?#1 (cast<int32x128> ?x0))",
    ),
    (
        180,
        2,
        "(get ?#1 (cast<unknown> (+ (cast<unknownx128> (call[max] (cast<unknownx128> (call[min] (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?#2) (cast<int32> 128))) (cast<int32> ?#3))) (cast<int32> (neg (cast<int32> 1))))) (cast<int32> 1) (cast<int32> 128))) (cast<int32x128> (call[x128] (cast<int32> (+ (cast<int32> ?#4) (cast<int32> (neg (cast<int32> 1))))))))) (cast<int32x128> (call[x128] (cast<int32> 0))))) (cast<int32x128> (call[x128] (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?#5))))))))",
    ),
    (
        183,
        2,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 256))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        185,
        2,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 258))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        186,
        2,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 384))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        188,
        2,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> 386))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        10,
        3,
        "(+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?#1) (cast<int32> 4))))) (cast<int32> 5))",
    ),
    (
        11,
        3,
        "(+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?#1) (cast<int32> 4))))) (cast<int32> 6))",
    ),
    (
        29,
        3,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (- (cast<int32> 0) (cast<int32> (* (cast<int32> ?#2) (cast<int32> ?#3))))))",
    ),
    (
        40,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 1))) (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 6))))) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        41,
        3,
        "(+ (cast<int32> (call[min] (cast<int32> (call[max] (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 1))) (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> 6))))) (cast<int32> ?x2))",
    ),
    (
        49,
        3,
        "(+ (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        95,
        3,
        "(+ (cast<unknown> ?x0) (cast<int32> 1))",
    ),
    (
        97,
        3,
        "(+ (cast<unknown> ?x0) (cast<int32> 128))",
    ),
    (
        99,
        3,
        "(+ (cast<unknown> ?x0) (cast<int32> ?x1))",
    ),
    (
        100,
        3,
        "(+ (cast<unknown> ?x0) (cast<unknown> ?x1))",
    ),
    (
        130,
        3,
        "(- (cast<int32> ?x0) (cast<int32> (* (cast<int32> ?x1) (cast<int32> 4))))",
    ),
    (
        154,
        3,
        "(* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> ?x0))",
    ),
    (
        173,
        3,
        "(<= (cast<int32> (call[max] (cast<int32> (call[min] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))) (cast<int32> ?x3))) (cast<int32> 1))) (cast<int32> ?x3))",
    ),
    (
        191,
        3,
        "(get ?#1 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?#2) (cast<int32> 128))) (cast<int32> ?x0))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        54,
        4,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> 1))))",
    ),
    (
        62,
        4,
        "(+ (cast<uint16x128> (call[absd] (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#16 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x2) (cast<uint8x128> ?x3))))) (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#17 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x4) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x6) (cast<uint8x128> ?x7))))))) (cast<uint16x128> (call[absd] (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#18 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x9))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x2) (cast<uint8x128> ?x6))))) (cast<uint16x128> (+ (cast<uint16x128> (call[widening_shift_left] (cast<uint8x128> (get ?#19 (cast<unknownx128> (call[ramp] (cast<unknown> (+ (cast<unknown> (* (cast<unknown> ?x8) (cast<int32> 128))) (cast<int32> ?x10))) (cast<int32> 1) (cast<int32> 128))))) (cast<uint8x128> (call[x128] (cast<uint8> 1))))) (cast<uint16x128> (call[widening_add] (cast<uint8x128> ?x3) (cast<uint8x128> ?x7))))))))",
    ),
    (
        88,
        4,
        "(+ (cast<unknown> (* (cast<unknown> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        149,
        5,
        "(* (cast<int32> ?x0) (cast<int32> 4))",
    ),
    (
        161,
        5,
        "(/ (cast<int32> (+ (cast<int32> (call[min] (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?x1))) ?x2)) (cast<int32> (neg (cast<int32> 127))))) (cast<int32> 128))",
    ),
    (
        148,
        6,
        "(* (cast<int32> (+ (cast<int32> ?#1) (cast<int32> ?x0))) (cast<int32> 4))",
    ),
    (
        25,
        8,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?x1))",
    ),
    (
        59,
        8,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        150,
        8,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        76,
        9,
        "(+ (cast<unknown> (* (cast<unknown> (/ (cast<unknown> (+ (cast<unknown> ?#1) (cast<int32> 127))) (cast<int32> 128))) (cast<int32> ?x0))) (cast<unknown> ?#2))",
    ),
    (
        164,
        9,
        "(/ (cast<int32> (- (cast<int32> 0) (cast<int32> ?x0))) (cast<int32> 128))",
    ),
    (
        169,
        11,
        "(/ ?x0 (cast<int32> 128))",
    ),
    (
        0,
        12,
        "(neg (cast<int32> ?x0))",
    ),
    (
        165,
        12,
        "(/ (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        53,
        13,
        "(+ (cast<int32> ?x0) (cast<int32> 127))",
    ),
    (
        96,
        13,
        "(+ (cast<unknown> ?x0) (cast<int32> 127))",
    ),
]
